src/stan/mcmc/chainset.hpp

```diff
@@ -0,0 +1,556 @@
+#ifndef STAN_MCMC_CHAINSET_HPP
+#define STAN_MCMC_CHAINSET_HPP
+
+#include <stan/io/stan_csv_reader.hpp>
+#include <stan/math/prim.hpp>
+#include <stan/analyze/mcmc/split_rank_normalized_ess.hpp>
+#include <stan/analyze/mcmc/split_rank_normalized_rhat.hpp>
+#include <boost/accumulators/accumulators.hpp>
+#include <boost/accumulators/statistics/stats.hpp>
+#include <boost/accumulators/statistics/mean.hpp>
+#include <boost/accumulators/statistics/tail_quantile.hpp>
+#include <boost/accumulators/statistics/p_square_quantile.hpp>
+#include <boost/accumulators/statistics/variance.hpp>
+#include <boost/accumulators/statistics/covariance.hpp>
+#include <boost/accumulators/statistics/variates/covariate.hpp>
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <sstream>
+#include <utility>
+#include <vector>
+#include <cstdlib>
+
+namespace stan {
+namespace mcmc {
+using Eigen::Dynamic;
+
+/**
+ * An <code>mcmc::chainset</code> object manages the post-warmup draws
+ * across a set of MCMC chains, which all have the same number or samples.
+ *
+ * <p><b>Storage Order</b>: Storage is column/last-index major.
+ */
+class chainset {
+ private:
+  size_t num_samples_;
+  std::vector<std::string> param_names_;
+  std::vector<Eigen::MatrixXd> chains_;
+
+  static size_t thinned_samples(const stan::io::stan_csv& stan_csv) {
+    size_t thinned_samples = stan_csv.metadata.num_samples;
+    if (stan_csv.metadata.thin > 0) {
+      thinned_samples = thinned_samples / stan_csv.metadata.thin;
+    }
+    return thinned_samples;
+  }
+
+  static bool is_valid(const stan::io::stan_csv& stan_csv) {
+    if (stan_csv.header.empty())
+      return false;
+    if (stan_csv.samples.size() == 0)
+      return false;
+    if (stan_csv.samples.rows() != thinned_samples(stan_csv))
+      return false;
+    return true;
+  }
```

Comment:
If these are static can they exist outside of the class? I think these are fine/good to expose as utility functions

Reply:
done.

Reply:
fixed in PR https://github.com/stan-dev/stan/pull/3310